picoquic/tls_api.c

```diff
@@ -429,6 +429,55 @@ uint64_t picoquic_get_simulated_time_cb(ptls_get_time_t* self)
     return ((**pp_simulated_time) / 1000);
 }
 
+typedef struct {
+    ptls_verify_certificate_t cb;
+    picoquic_quic_t *quic;
+} picoquic_verify_certificate_t;
+
+typedef struct {
+    /* The pointer to the overlying `verify_ctx` */
+    void *verify_ctx;
+    int (*verify_sign)(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign);
+} picoquic_verify_ctx_t;
+
+static int verify_sign_callback(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign)
+{
+    picoquic_verify_ctx_t* ctx = (picoquic_verify_ctx_t*)verify_ctx;
+    int ret = 0;
+
+    ret = ctx->verify_sign(ctx->verify_ctx, data, sign);
+
+    free(ctx);
+
+    return ret;
+}
+
+#define container_of(ptr, type, member) ((type *)((char *)(ptr) - offsetof(type, member)))
+
+static int verify_certificate_callback(ptls_verify_certificate_t* _self, void* ctx, ptls_t *tls,
+                                       int (**verify_sign)(void *verify_ctx, ptls_iovec_t data, ptls_iovec_t sign),
+                                       void **verify_data,
+                                       ptls_iovec_t *certs,
+                                       size_t num_certs)
+{
+    picoquic_verify_certificate_t *self = container_of(_self, picoquic_verify_certificate_t, cb);
+    picoquic_cnx_t* cnx = (picoquic_cnx_t*)ctx;
```

Comment:
The macro "container of" is Linux specific, and will not compile in Windows. For portability, you want something like:
picoquic_verify_certificate_t *self = (picoquic_verify_certificate_t *)((char*)_self - offsetof(struct st_picoquic_verify_certificate_t , cb));

Reply:
I defined the macro in line 455 :D

Reply:
Sorry, missed that.